src/Servers/Kestrel/Core/src/Internal/Http2/Http2OutputProducer.cs

```diff
@@ -66,20 +70,103 @@ public Http2OutputProducer(Http2Stream stream, Http2StreamContext context, Strea
         _flusher = new TimingPipeFlusher(timeoutControl: null, _log);
         _flusher.Initialize(_pipeWriter);
+        _window = context.ClientPeerSettings.InitialWindowSize;
+    }
+
+    public Http2Stream Stream => _stream;
+    public PipeReader PipeReader => _pipeReader;
+
+    public bool IsTimingWrite { get; set; }
+
+    public bool WriteHeaders { get; set; }
+
+    public bool StreamEnded => _streamEnded;
+
+    public bool StreamCompleted
+    {
+        get
+        {
+            lock (_dataWriterLock)
+            {
+                return _streamCompleted;
+            }
+        }
+    }
+
+    // Useful for debugging the scheduling state in the debugger
+    internal (int, long, State, long, bool) SchedulingState => (Stream.StreamId, _unconsumedBytes, _observationState, _window, _waitingForWindowUpdates);
+
+    // Added bytes to the queue.
+    // Returns a bool that represents whether we should schedule this producer to write
+    // the enqueued bytes
+    private bool Enqueue(long bytes)
+    {
+        lock (_dataWriterLock)
+        {
+            var wasEmpty = _unconsumedBytes == 0;
+            _unconsumedBytes += bytes;
+            return wasEmpty && _unconsumedBytes > 0 && _observationState == State.None;
+        }
+    }
+
+    // Determines if we should schedule this producer to observe
+    // any state changes made.
+    private bool EnqueueForObservation(State state)
+    {
+        lock (_dataWriterLock)
+        {
+            var wasEnqueuedForObservation = _observationState != State.None;
+            _observationState |= state;
+            return (_unconsumedBytes == 0 || _waitingForWindowUpdates) && !wasEnqueuedForObservation;
+        }
+    }
+
+    // Removes consumed bytes from the queue.
+    // Returns a bool that represents whether we should schedule this producer to write
+    // the remaining bytes.
+    internal (bool, bool) Dequeue(long bytes, State state)
+    {
+        lock (_dataWriterLock)
+        {
+            _observationState &= ~state;
+            _unconsumedBytes -= bytes;
+            return (_unconsumedBytes > 0, _observationState != State.None);
+        }
+    }
+
+    // Consumes bytes from the stream's window and returns the remaining bytes and actual bytes consumed
+    internal (long, long) ConsumeWindow(long bytes)
+    {
+        lock (_dataWriterLock)
+        {
+            var actual = Math.Min(bytes, _window);
+            var remaining = _window -= actual;
+            return (actual, remaining);
+        }
+    }
 
-        _dataWriteProcessingTask = ProcessDataWrites();
+    // Adds more bytes to the stream's window
+    // Returns a bool that represents whether we should schedule this producer to write
+    // the remaining bytes.
+    private bool UpdateWindow(long bytes)
+    {
+        lock (_dataWriterLock)
+        {
+            var wasDepleted = _window <= 0;
```

Comment:
Can `_window` ever go below 0?

Reply:
Yes, guess how I found out... The tests!